packages/network-controller/src/NetworkController.test.ts

```diff
@@ -1,278 +1,4156 @@
+        expect(controller.state).toStrictEqual({
+          network: 'loading',
+          isCustomNetwork: false,
+          providerConfig: { type: 'mainnet' as const, chainId: '1' },
+          properties: { isEIP1559Compatible: false },
+        });
+      });
     });
 
-    expect(controller.state).toStrictEqual({
-      network: 'loading',
-      isCustomNetwork: false,
-      properties: { isEIP1559Compatible: false },
-      providerConfig: {
-        type: 'mainnet',
-        chainId: '1',
-      },
+    it('merges the given state into the default state', async () => {
+      await withController(
+        {
+          state: {
+            isCustomNetwork: true,
+            properties: { isEIP1559Compatible: true },
+          },
+        },
+        ({ controller }) => {
+          expect(controller.state).toStrictEqual({
+            network: 'loading',
+            isCustomNetwork: true,
+            providerConfig: { type: 'mainnet', chainId: '1' },
+            properties: { isEIP1559Compatible: true },
+          });
+        },
+      );
     });
   });
 
-  it('should create a provider instance for default infura network', () => {
-    const networkControllerOpts = {
-      infuraProjectId: 'foo',
-      messenger,
-    };
-    const controller = new NetworkController(networkControllerOpts);
-    controller.providerConfig = {} as ProviderConfig;
-    expect(controller.provider instanceof Web3ProviderEngine).toBe(true);
-  });
+  describe('providerConfig property', () => {
+    describe('get', () => {
+      it('throws', async () => {
+        await withController(({ controller }) => {
+          expect(() => controller.providerConfig).toThrow(
+            'Property only used for setting',
+          );
+        });
+      });
+    });
+
+    describe('set', () => {
+      ['1', '5', '11155111', ''].forEach((chainId) => {
+        describe(`when the provider config in state contains a chain ID of "${chainId}"`, () => {
+          it('sets isCustomNetwork in state to false', async () => {
+            await withController(
+              {
+                state: {
+                  isCustomNetwork: true,
+                  providerConfig: buildProviderConfig({
+                    chainId,
+                  }),
+                },
+                infuraProjectId: 'infura-project-id',
+              },
+              ({ controller }) => {
+                const fakeInfuraProvider = buildFakeInfuraProvider();
+                createInfuraProviderMock.mockReturnValue(fakeInfuraProvider);
+                const fakeInfuraSubprovider = buildFakeInfuraSubprovider();
+                SubproviderMock.mockReturnValue(fakeInfuraSubprovider);
+                const fakeMetamaskProvider = buildFakeMetamaskProvider();
+                createMetamaskProviderMock.mockReturnValue(
+                  fakeMetamaskProvider,
+                );
+
+                controller.providerConfig = buildProviderConfig();
```

Comment:
This test and the following test are both formatted as though they are testing the effects of setting `buildProviderConfig();`. But it's not; this line has no impact upon the assertion, because we're not waiting for the initialization triggered by this setter to complete.

The assertions in both of these tests are just validating the effects of passing in a particular chainId to the constructor.

Reply:
Oh interesting. So the `providerConfig` passed to the `providerConfig` setter is not used in practice. Or rather, the properties on the `ProviderConfig` type are not used in practice (they're stored but overwritten). Additional properties that go beyond the typed properties are used though, in that they're passed into `web3-provider-engine`, which mobile depends upon.

Reply:
Effectively the config passed to this setter are saved on `this.internalProviderConfig`, which in practice is mixed in with `this.state.providerConfig` with the state overriding this internal config. This combined state is passed to `createMetamaskProvider`, so properties beyond those allowed by the type are passed through and used by `createMetamaskProvider`

Reply:
Yes, I think you've assessed this correctly. It's not straightforward.

Reply:
Updated the names of these tests to clarify this: [22e62ec](https://github.com/MetaMask/core/pull/1039/commits/22e62ecc982b81f52d30546ec37b777c631a14bd)

Reply:
Thanks, that looks good